File: src/test/java/io/github/dmitrikudrenko/calculator/CalculatorScreen.java
package io.github.dmitrikudrenko.calculator;

import io.appium.java_client.AppiumDriver;
import io.appium.java_client.MobileElement;
import io.appium.java_client.pagefactory.AndroidFindBy;
import io.github.dmitrikudrenko.core.AbstractScreen;
import org.openqa.selenium.TimeoutException;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

public class CalculatorScreen extends AbstractScreen {
    @AndroidFindBy(id = CalculatorConstants.ID + "digit_0")
    private MobileElement zeroBtn;

    @AndroidFindBy(id = CalculatorConstants.ID + "digit_1")
    private MobileElement oneBtn;

    @AndroidFindBy(id = CalculatorConstants.ID + "digit_2")
    private MobileElement twoBtn;

    @AndroidFindBy(id = CalculatorConstants.ID + "digit_3")
    private MobileElement threeBtn;

    @AndroidFindBy(id = CalculatorConstants.ID + "digit_4")
    private MobileElement fourBtn;

    @AndroidFindBy(id = CalculatorConstants.ID + "digit_5")
    private MobileElement fiveBtn;

    @AndroidFindBy(id = CalculatorConstants.ID + "digit_6")
    private MobileElement sixBtn;

    @AndroidFindBy(id = CalculatorConstants.ID + "digit_7")
    private MobileElement sevenBtn;

    @AndroidFindBy(id = CalculatorConstants.ID + "digit_8")
    private MobileElement eightBtn;

    @AndroidFindBy(id = CalculatorConstants.ID + "digit_9")
    private MobileElement nineBtn;

    @AndroidFindBy(id = CalculatorConstants.ID + "op_add")
    private MobileElement plusBtn;

    @AndroidFindBy(id = CalculatorConstants.ID + "op_sub")
    private MobileElement minusBtn;

    @AndroidFindBy(id = CalculatorConstants.ID + "op_div")
    private MobileElement divideBtn;

    @AndroidFindBy(id = CalculatorConstants.ID + "op_mul")
    private MobileElement multipleBtn;

    @AndroidFindBy(id = CalculatorConstants.ID + "eq")
    private MobileElement equalsBtn;

    @AndroidFindBy(id = CalculatorConstants.ID + "dec_point")
    private MobileElement decimalPointBtn;

    @AndroidFindBy(id = CalculatorConstants.ID + "result")
    private MobileElement resultField;

    @AndroidFindBy(id = CalculatorConstants.ID + "formula")
    private MobileElement formulaField;

    public CalculatorScreen(AppiumDriver driver) {
        super(driver);
    }

    public String getResult(String result) {
        try {
            (new WebDriverWait(driver, 5))
                    .until(ExpectedConditions.textToBePresentInElement(resultField, result));
        } catch (TimeoutException ignored) {
        }
        return resultField.getText();
    }

    public MobileElement getZeroBtn() {
        return zeroBtn;
    }

    public MobileElement getOneBtn() {
        return oneBtn;
    }

    public MobileElement getTwoBtn() {
        return twoBtn;
    }

    public MobileElement getThreeBtn() {
        return threeBtn;
    }

    public MobileElement getFourBtn() {
        return fourBtn;
    }

    public MobileElement getFiveBtn() {
        return fiveBtn;
    }

    public MobileElement getSixBtn() {
        return sixBtn;
    }

    public MobileElement getSevenBtn() {
        return sevenBtn;
    }

    public MobileElement getEightBtn() {
        return eightBtn;
    }

    public MobileElement getNineBtn() {
        return nineBtn;
    }

    public MobileElement getPlusBtn() {
        return plusBtn;
    }

    public MobileElement getMinusBtn() {
        return minusBtn;
    }

    public MobileElement getDivideBtn() {
        return divideBtn;
    }

    public MobileElement getMultipleBtn() {
        return multipleBtn;
    }

    public MobileElement getEqualsBtn() {
        return equalsBtn;
    }

    public MobileElement getDecimalPointBtn() {
        return decimalPointBtn;
    }
}

File: src/test/kotlin/io/github/dmitrikudrenko/calculator/core/CalculatorScreen.kt
package io.github.dmitrikudrenko.calculator.core

import io.appium.java_client.AppiumDriver
import io.appium.java_client.MobileElement
import io.github.dmitrikudrenko.core.AbstractScreen
import io.github.dmitrikudrenko.core.bindOptionalElement
import io.github.dmitrikudrenko.core.shouldClick
import io.github.dmitrikudrenko.core.shouldHaveText

class CalculatorScreen(driver: AppiumDriver<*>) : AbstractScreen(driver) {
    val zeroBtn: MobileElement? by bindOptionalElement(constants.id + "digit_0")

    val oneBtn: MobileElement? by bindOptionalElement(constants.id + "digit_1")

    val twoBtn: MobileElement? by bindOptionalElement(constants.id + "digit_2")

    val threeBtn: MobileElement? by bindOptionalElement(constants.id + "digit_3")

    val fourBtn: MobileElement? by bindOptionalElement(constants.id + "digit_4")

    val fiveBtn: MobileElement? by bindOptionalElement(constants.id + "digit_5")

    val sixBtn: MobileElement? by bindOptionalElement(constants.id + "digit_6")

    val sevenBtn: MobileElement? by bindOptionalElement(constants.id + "digit_7")

    val eightBtn: MobileElement? by bindOptionalElement(constants.id + "digit_8")

    val nineBtn: MobileElement? by bindOptionalElement(constants.id + "digit_9")

    val plusBtn: MobileElement? by bindOptionalElement(constants.id + "op_add")

    val minusBtn: MobileElement? by bindOptionalElement(constants.id + "op_sub")

    val divideBtn: MobileElement? by bindOptionalElement(constants.id + "op_div")

    val multipleBtn: MobileElement? by bindOptionalElement(constants.id + "op_mul")

    val equalsBtn: MobileElement? by bindOptionalElement(constants.id + "eq")

    val decimalPointBtn: MobileElement? by bindOptionalElement(constants.id + "dec_point")

    val deleteBtn: MobileElement? by bindOptionalElement(constants.id + "del")

    val clearBtn: MobileElement? by bindOptionalElement(constants.id + "clr")

    val resultField: MobileElement? by bindOptionalElement(constants.id + "result")

    val formulaField: MobileElement? by bindOptionalElement(constants.id + "formula")

    fun inputFormula(formula: String): CalculatorScreen {
        for (digit in formula.toCharArray()) {
            when (digit) {
                '0' -> zeroBtn.shouldClick()
                '1' -> oneBtn.shouldClick()
                '2' -> twoBtn.shouldClick()
                '3' -> threeBtn.shouldClick()
                '4' -> fourBtn.shouldClick()
                '5' -> fiveBtn.shouldClick()
                '6' -> sixBtn.shouldClick()
                '7' -> sevenBtn.shouldClick()
                '8' -> eightBtn.shouldClick()
                '9' -> nineBtn.shouldClick()
                '+' -> plusBtn.shouldClick()
                '−' -> minusBtn.shouldClick()
                '×' -> multipleBtn.shouldClick()
                '÷' -> divideBtn.shouldClick()
                '=' -> equalsBtn.shouldClick()
            }
        }
        return this
    }

    fun shouldBeResult(result: String): CalculatorScreen {
        resultField.shouldHaveText(result)
        return this
    }
}
